core/mm/file.c

```diff
@@ -122,9 +122,32 @@ struct file *file_get_by_tag(const uint8_t *tag, unsigned int taglen)
 		return NULL;
 
 	mutex_lock(&file_mu);
-	f = file_get(file_find_tag_unlocked(tag, taglen));
-	if (f)
+
+	/*
+	 * If file is found and reference counter can be increased, we're done.
+	 * If file can't be found, it doesn't exist so it has to be added.
+	 * If it's found but reference counter is 0, the situation is
+	 * a bit complicated:
+	 * - file_put() is about to free the file as soon as it can obtain the
+	 *   mutex.
+	 * - Unless there's a mismatch between file_get() and file_put() only
+	 *   one thread calling file_put() is about to free the file.
+	 *
+	 * There's a window of oportunity where file_put() is called
+	 * (without a mutex being held, which is quite OK) while we're
+	 * holding the mutex here and are searching for the file and it's
+	 * found, but just after file_put() has decreased the reference
+	 * counter.
```

Comment:
Wouldn't it be much simpler to always grab the mutex before touching anything?...

Reply:
Yes it would, but then we'd have more congestion too.

Reply:
Maybe, but I'd be curious about the performance impact. Anyway, this looks OK, although I'm still trying to convince myself there can be no other race condition...

Reply:
I have no numbers on performance impact.